apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl

```diff
@@ -0,0 +1,372 @@
+-define(shard, 2).
+-define(enc, 3).
+
+-define(IT, 61).
+-define(DELETE_IT, 62).
+
+-type shard() :: binary().
+
+-opaque iterator() ::
+    #{
+        ?tag := ?IT,
+        ?shard := shard(),
+        ?enc := term()
+    }.
+
+-opaque delete_iterator() ::
+    #{
+        ?tag := ?DELETE_IT,
+        ?shard := shard(),
+        ?enc := term()
+    }.
+
+-type db_opts() ::
+    #{
+        backend := builtin_local,
+        storage := emqx_ds_storage_layer:prototype(),
+        n_shards := pos_integer()
+    }.
+
+-type generation_rank() :: {shard(), emqx_ds_storage_layer:gen_id()}.
+
+-define(stream(SHARD, INNER), [2, SHARD | INNER]).
+-define(delete_stream(SHARD, INNER), [3, SHARD | INNER]).
+
+%%================================================================================
+%% API functions
+%%================================================================================
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+-spec open_db(emqx_ds:db(), db_opts()) -> ok | {error, _}.
+open_db(DB, CreateOpts) ->
+    case emqx_ds_builtin_local_sup:start_db(DB, CreateOpts) of
+        {ok, _} ->
+            ok;
+        {error, {already_started, _}} ->
+            ok;
+        {error, Err} ->
+            {error, Err}
+    end.
+
+-spec close_db(emqx_ds:db()) -> ok.
+close_db(DB) ->
+    emqx_ds_builtin_local_sup:stop_db(DB).
+
+-spec add_generation(emqx_ds:db()) -> ok | {error, _}.
+add_generation(DB) ->
+    Shards = emqx_ds_builtin_local_meta:shards(DB),
+    Errors = lists:filtermap(
+        fun(Shard) ->
+            ShardId = {DB, Shard},
+            case
+                emqx_ds_storage_layer:add_generation(
+                    ShardId, emqx_ds_builtin_local_meta:ensure_monotonic_timestamp(ShardId)
+                )
+            of
+                ok ->
+                    false;
+                Error ->
+                    {true, {Shard, Error}}
+            end
+        end,
+        Shards
+    ),
+    case Errors of
+        [] -> ok;
+        _ -> {error, Errors}
+    end.
+
+-spec update_db_config(emqx_ds:db(), db_opts()) -> ok | {error, _}.
+update_db_config(DB, CreateOpts) ->
+    Opts = #{} = emqx_ds_builtin_local_meta:update_db_config(DB, CreateOpts),
+    lists:foreach(
+        fun(Shard) ->
+            ShardId = {DB, Shard},
+            emqx_ds_storage_layer:update_config(
+                ShardId, emqx_ds_builtin_local_meta:ensure_monotonic_timestamp(ShardId), Opts
+            )
+        end,
+        emqx_ds_builtin_local_meta:shards(DB)
+    ).
+
+-spec list_generations_with_lifetimes(emqx_ds:db()) ->
+    #{emqx_ds:generation_rank() => emqx_ds:generation_info()}.
+list_generations_with_lifetimes(DB) ->
+    lists:foldl(
+        fun(Shard, Acc) ->
+            maps:fold(
+                fun(GenId, Data, Acc1) ->
+                    Acc1#{{Shard, GenId} => Data}
```

Comment:
Since `emqx_persistent_message_ds_gc_worker` currently expects lifetime-related timestamps are in milliseconds, this will break those expectations AFAICT. Probably either timestamps need to be converted back, or callers taught to deal with finer timestamps?

Reply:
Good catch, fixed.
P.S. That was one of the reasons I was against unit conversions in DS :slightly_smiling_face: